storage/filesystem/mmap/fsobject.go

```diff
@@ -0,0 +1,343 @@
+	if o.diskType.IsDelta() {
+		data := o.scanner.packMmap[start : start+o.size]
+		br := sync.GetBufioReader(bytes.NewReader(data))
+		return io.NopCloser(br), nil
+	}
+
+	data := o.scanner.packMmap[start:]
+	br := sync.GetBufioReader(bytes.NewReader(data))
+	zr, err := sync.GetZlibReader(br)
+	if err != nil {
+		return nil, err
+	}
+	return &zlibReadCloser{r: zr, rbuf: br}, nil
+}
+
+// Hash holds the object's ID.
+func (o *ondemandObject) Hash() plumbing.Hash {
+	return o.hash
+}
+
+// Size holds the object's size.
+func (o *ondemandObject) Size() int64 {
+	o.m.RLock()
+	defer o.m.RUnlock()
+
+	return o.size
+}
+
+// Type holds the object's ObjectType.
+func (o *ondemandObject) Type() plumbing.ObjectType {
+	o.m.RLock()
+	defer o.m.RUnlock()
+
+	return o.typ
+}
+
+// TODO: Create a read-only EncodedObject interface, to avoid the LSP
+// violation as per methods below.
+
+// SetSize only exists to implement the plumbing.EncodedObject interface.
+// This method has no effect to the underlying, as it is a no-op.
+func (o *ondemandObject) SetSize(int64) {}
+
+// SetType only exists to implement the plumbing.EncodedObject interface.
+// This method has no effect to the underlying, as it is a no-op.
+func (o *ondemandObject) SetType(plumbing.ObjectType) {}
+
+// Writer only exists to implement the plumbing.EncodedObject interface.
+// This method always returns a nil writer.
+func (o *ondemandObject) Writer() (io.WriteCloser, error) {
+	return nil, nil
+}
+
+// resolveMetadata resolves the type and size for the delta object without
+// fully materializing it.
+//
+// Calling it on a non-delta object is a no-op. Subsequent calls on a delta
+// object will also become a no-op.
+func (o *ondemandObject) resolveMetadata() error {
+	o.m.RLock()
+	if !o.typ.IsDelta() {
+		o.m.RUnlock()
+		return nil
+	}
+	o.m.RUnlock()
+
+	o.m.Lock()
+	defer o.m.Unlock()
+
+	// enable Reader() to get the resolved content, as opposed to delta's.
+	o.autoResolve = true
+
+	pos := o.toDataOffset(o.offset)
+
+	var baseOffset uint64
+	var baseHash plumbing.Hash
+	var err error
+
+	if o.diskType == plumbing.OFSDeltaObject {
+		reader := bytes.NewReader(o.scanner.packMmap[pos:])
+		negativeOffset, err := binary.ReadVariableWidthInt(reader)
+		if err != nil {
+			return fmt.Errorf("failed to read OFS delta offset: %w", err)
+		}
+
+		baseOffset = uint64(o.offset) - uint64(negativeOffset)
+		consumed := len(o.scanner.packMmap[pos:]) - reader.Len()
+		pos += int64(consumed)
+	} else {
+		hashSize := o.scanner.hashSize
+		if pos+int64(hashSize) > int64(len(o.scanner.packMmap)) {
+			return fmt.Errorf("invalid REF delta: hash out of bounds")
+		}
+
+		baseHash, _ = plumbing.FromBytes(o.scanner.packMmap[pos : pos+int64(hashSize)])
+		pos += int64(hashSize)
+	}
+
+	// Get the base object to inherit its type.
+	var base plumbing.EncodedObject
+	if o.diskType == plumbing.OFSDeltaObject {
+		base, err = o.scanner.GetByOffset(baseOffset)
+	} else {
+		base, err = o.scanner.Get(baseHash)
+	}
```

Comment:
```suggestion
	// Get the base object to inherit its type.
	var base plumbing.EncodedObject
	var err error

	if o.diskType == plumbing.OFSDeltaObject {
		reader := bytes.NewReader(o.scanner.packMmap[pos:])
		negativeOffset, err := binary.ReadVariableWidthInt(reader)
		if err != nil {
			return fmt.Errorf("failed to read OFS delta offset: %w", err)
		}

		baseOffset := uint64(o.offset) - uint64(negativeOffset)
		consumed := len(o.scanner.packMmap[pos:]) - reader.Len()
		pos += int64(consumed)
		
		base, err = o.scanner.GetByOffset(baseOffset)
	} else {
		hashSize := o.scanner.hashSize
		if pos+int64(hashSize) > int64(len(o.scanner.packMmap)) {
			return fmt.Errorf("invalid REF delta: hash out of bounds")
		}

		baseHash, _ := plumbing.FromBytes(o.scanner.packMmap[pos : pos+int64(hashSize)])
		pos += int64(hashSize)
		
		base, err = o.scanner.Get(baseHash)
	}
```

IMO it looks cleaner to limit the scope of `baseOffset` and `baseHash`.

Reply:
Nice one, completely agree. :+1: